### PR TITLE
[FIX] range: correctly handle unbounded ranges on row/col changes

### PR DESCRIPTION
**Pull request by rmbh-odoo**
## Description:

Steps to reproduce:
- Create a sheet with 3 rows
- Set `A1 = 1`, `A2 = 2`, and `C1 = SUM(A:A)`
- Delete row 3
- Add a new row at the bottom
- Set `A3 = 6`

Issue:
- Unbounded ranges (e.g. `A:A`) are not properly updated on row/col changes.
  The changeType can be incorrectly computed as `NONE` or `MOVE` instead
  of `RESIZE`, causing formulas to miss newly added cells (e.g. `A3` in `SUM`).

Cause:
- Incomplete condition handling for unbounded ranges caused inconsistent
  changeType detection across different scenarios.

Fix:
- Refine changeType computation for unbounded ranges by covering all edge
  cases and ensuring correct `RESIZE` behavior when the range should expand.

Task: [6167358](https://www.odoo.com/odoo/2328/tasks/6167358)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/helpers/range.ts [hunk 1]
```diff
@@ -59,21 +59,29 @@ interface RangeXcArgs {
...
     parts = [args.parts[0], args.parts[0]];
-  } else if (args.parts.length === 2 && getZoneArea(zone) === 1) {
+  } else if (args.parts.length === 2 && getZoneArea(initialZone) === 1) {
     parts = [args.parts[0]];
   }
-  return {
+  const range = {
     unboundedZone,
-    zone,
     parts,
```
Comment (rmbh-odoo) [inline src/helpers/range.ts:72]: I'm not entirely sure whether `parts` should also be dynamic. Since `parts` are derived from the `zone`, they logically should update when the `zone` changes.
However, at the moment, I’m not observing any side effects or issues with `parts` being static.

File: src/helpers/range.ts [hunk 2]
```diff
@@ -59,21 +59,29 @@ interface RangeXcArgs {
...
     prefixSheet: args.prefixSheet,
     invalidSheetName: args.invalidSheetName,
     sheetId: args.sheetId,
-  };
+  } as Range;
+
+  Object.defineProperty(range, "zone", {
```
Comment (dhrp-odoo) [inline src/helpers/range.ts]: I like that idea but there are still few bugs.

`createRange()` now makes `range.zone` dynamic, but `createAdaptedRange()` still rebuilds ranges with object spread (and there are other places as well)

To reproduce here are the steps:

- Start with a sheet that has 3 rows.
- Put =1 in A1.
- Put =2 in A2.
- Put =SUM(A:A) in C1.
- Delete row 3.
- Add 1 row at the bottom.
- Put =6 in A3.

**Review by dhrp-odoo — COMMENTED** [review]
👋

File: src/helpers/range.ts (hunk 2 above)
Comment (rmbh-odoo) [reply src/helpers/range.ts]: I've now introduced a cloneRange helper and replaced object spread usage, so the zone getter is preserved.

File: src/helpers/range.ts [hunk 3]
```diff
@@ -178,7 +186,7 @@ export function duplicateRangeInDuplicatedSheet(
   range: Range
 ): Range {
   const sheetId = range.sheetId === sheetIdFrom ? sheetIdTo : range.sheetId;
-  return { ...range, sheetId };
+  return cloneRange(range, { sheetId });
```
Comment (hokolomopo) [inline src/helpers/range.ts]: So maybe switching from a `Range` class to a POJO was not a good idea ... We're back at the start with `clone` methods, but now we aren't type-safe, and nothing prevents us from spreading a range to copy it ...

cc @LucasLefevre

File: tests/range_plugin.test.ts [hunk 4]
```diff
@@ -667,6 +667,20 @@ describe("full column range", () => {
     addRows(m, "before", 0, 1);
     expect(m.getters.getUsedRanges()[1]).toEqual("B2:C");
   });
+  test("full-column range updates correctly on row insertions and deletions", () => {
```
Comment (hokolomopo) [inline tests/range_plugin.test.ts]:
```suggestion
});

  test("full-column range updates correctly on row insertions and deletions", () => {
```
nitpick. same for the other test

**Review by hokolomopo — COMMENTED** [review]
@LucasLefevre

File: src/helpers/range.ts (hunk 2 above)
Comment (LucasLefevre) [inline src/helpers/range.ts]: This is magic, like it was before with the `Range` facade and `RangeImpl`.
Code shouldn't be magic 😉 

The POJO has no magic at all... at the cost of having two sources of truth (`unboundedZone` and `zone`) that can lead to bugs such as this one.

I'd much rather prefer to fix it with something like this 👇 easy, more explicit, no magic.

(other cases needs to be fixed in similar ways)

```patch
diff --git a/src/helpers/range.ts b/src/helpers/range.ts
index bd37d1321..8a2c2b762 100644
--- a/src/helpers/range.ts
+++ b/src/helpers/range.ts
@@ -438,7 +438,7 @@ function getApplyRangeChangeAddColRow(cmd: AddColumnsRowsCommand): ApplyRangeCha
       return { changeType: "NONE", range };
     }
     if (cmd.position === "after") {
-      if (range.zone[start] <= cmd.base && cmd.base < range.zone[end]) {
+      if (range.zone[start] <= cmd.base && (cmd.base < range.zone[end] || range.unboundedZone[end] === undefined)) {
         return {
           changeType: "RESIZE",
           range: createAdaptedRange(range, dimension, "RESIZE", cmd.quantity),
```

File: src/helpers/range.ts (hunk 1 above)
Comment (rmbh-odoo) [reply src/helpers/range.ts:72]: We remove the dynamic zone and find another way to consider the POJO approach :smile:

File: tests/range_plugin.test.ts (hunk 4 above)
Comment (rmbh-odoo) [reply tests/range_plugin.test.ts]: In this describe block, none of the test cases have spacing, so I kept it consistent and didn’t add any new line :sob: 
Do you think it’s worth adding new lines for all test cases in a separate commit, wdyt?

Comment (LucasLefevre) [reply tests/range_plugin.test.ts]: Not worth it IMO ;)
Keep it consistent 👍

File: src/helpers/range.ts (hunk 2 above)
Comment (rmbh-odoo) [reply src/helpers/range.ts]: I think I’ve covered all possible edge cases, so the ranges shouldn’t break anymore :slightly_smiling_face:

**Comment by rrahir** [comment]
robodoo r+
